[WIP] Refactor exception handling of `Binary\Loader\ChainLoader` class

robfrawley opened the pull request:
| Q | A
| --- | ---
| Branch? | 3.x <!--choose version number and select target branch accordingly-->
| Bug fix? | no
| New feature? | no
| BC breaks? | no
| Deprecations? | no
| References | https://github.com/liip/LiipImagineBundle/pull/1432#discussion_r751265173 / https://github.com/liip/LiipImagineBundle/pull/1432#discussion_r753720665
| License | MIT
| Doc | n/a

This is a __work in progress__ toward refactoring the error handling logic of [`Liip\ImagineBundle\Binary\Loader\ChainLoader`](https://github.com/liip/LiipImagineBundle/blob/3.x/src/Binary/Loader/ChainLoader.php#L1) as was originally discussed in code review comments from #1432; see the discussion beginning with *[dbu's #1432 comment](https://github.com/liip/LiipImagineBundle/pull/1432#discussion_r751265173)* and, more specifically, *[my (robfrawley) #1432 comment](https://github.com/liip/LiipImagineBundle/pull/1432#discussion_r753720665)* in response. 

Additional information will be added *as my schedule permits* and *during this PR's progression*.

For the time being, I would *greatly appreciate* any thoughts from @dbu, @franmomu, or any other maintainer/contributor/user as to the current state of this pull request as well as any alternate possible directions to investigate.

franmomu commented on src/Binary/Loader/ChainLoader.php:
```diff
@@ -37,36 +39,14 @@ public function find($path)
         $exceptions = [];
 
-        foreach ($this->loaders as $loader) {
+        foreach ($this->loaders as $configName => $objectInst) {
```
```suggestion
        foreach ($this->loaders as $configName => $loader) {
```
Just a suggestion that keeping `$loader` as variable seems more understandable to me.

robfrawley replied:
Sounds good; done.

robfrawley replied:
Sounds good; done.

*(Note: I do agree that the use of `loader` (especially for user-facing code) is much clearer using the suggested (and original) variable name (instead of `objectInst`); in fact, I never actually intended to change that variable at all! ;-) I think the loader variable was altered when I was refactoring the new exception class properties and accidentally cast too wide a "find/replace" net, such that the ChainLoader::$loader property was mistakenly included in the change. Regardless, it's fixed now. I was working on a machine without my normal environment, PHPStorm, and was instead using SublimeText, so such unexpected changes shouldn't normally be an issue. Thanks for the catch, though.)*

franmomu commented on src/Exception/Binary/Loader/ChainAttemptNotLoadableException.php:
```diff
@@ -0,0 +1,53 @@
+    private string $configName;
+    private LoaderInterface $objectInst;
+
+    public function __construct(string $configName, LoaderInterface $objectInst, NotLoadableException $loaderException)
```
```suggestion
    public function __construct(string $configName, LoaderInterface $loader, NotLoadableException $loaderException)
```
same here

robfrawley replied:
@franmomu @dbu Are you both comfortable with changing `objectInst` to `loaderInst` (instead of the suggested variable name of `loader`) with regard to the *exception class __only__*? Specifically:

- `ChainAttemptNotLoadableException::$objectInst` *->* __the original property name__
- `ChainAttemptNotLoadableException::$loaderInst` *->* __my suggested property name__

This better aligns with the methods as they are currently defined for the chain-attempt exception, IMHO. For example:

- `ChainAttemptNotLoadableException::getLoaderConfigName()`
- `ChainAttemptNotLoadableException::getLoaderObjectInst()`
- `ChainAttemptNotLoadableException::getLoaderObjectName()`

*(Not that it would matter to anyone else, but my suggestion also better satisfies my OCD which prefers same-length groups of properties and methods, when feasible to do so without convoluting their meaning...)*

Let me know if this is objectionable to either of you; I'll perform the original change if you both are not agreeable to my suggested alternative.

franmomu commented on src/Exception/Binary/Loader/ChainAttemptNotLoadableException.php:
```diff
@@ -0,0 +1,53 @@
+
+use Liip\ImagineBundle\Binary\Loader\LoaderInterface;
+
+class ChainAttemptNotLoadableException extends NotLoadableException
```
```suggestion
final class ChainAttemptNotLoadableException extends NotLoadableException
```
I think we should make them `final` by default if it is not designed to be extended.

dbu replied:
i agree, makes it much more explicit to know for backwards compatibility

robfrawley replied:
Excellent idea; done.

franmomu reviewed (commented):
Nice refactor!

dbu commented:
i like this refactoring, thanks!

i agree with the comment from @franmomu about naming the loader variable $loader and making things final where possible.

robfrawley commented:
I've gone ahead and squashed the five commits created while working toward this PR, as well as rebased my changes over the latest `3.x` branch. I think this should be just about ready, pending the outcome of https://github.com/liip/LiipImagineBundle/pull/1434#discussion_r757095083 and any additional thoughts that anyone may have after performing any final reviews. As such, I've removed the `[WIP]` label from the PR title and I request a final quick check from any interested maintainers and contributors.

*(At this point, I'll squash and rebase every time any additional changes are required so the PR is always in a clean and easily mergeable state; hopefully such doesn't happen too many times, though :-))*

dbu commented on src/Exception/Binary/Loader/ChainAttemptNotLoadableException.php:
```diff
@@ -0,0 +1,53 @@
+        return $this->loaderInst;
+    }
+
+    public function getLoaderObjectName(): string
```
if i understand correctly, this is the class name without the namespace. how about calling the method `getLoaderClassName`?

dbu commented on src/Exception/Binary/Loader/ChainAttemptNotLoadableException.php:
```diff
@@ -0,0 +1,53 @@
+        return $this->configName;
+    }
+
+    public function getLoaderObjectInst(): LoaderInterface
```
yeah naming things... i would call the thing just `$loader` and this method `getLoader`. i prefer concise names. that it is an instantiated object is defined by the return type declaration.

dbu reviewed (commented):
i would prefer to do some renaming here as i feel both "object" and "inst" do not help understanding the code, so just "loader" would be my preferred solution.

take it as a suggestion, if you really want the current name, we can also keep it.

dbu commented on src/Exception/Binary/Loader/ChainAttemptNotLoadableException.php:
```diff
@@ -0,0 +1,53 @@
+final class ChainAttemptNotLoadableException extends NotLoadableException
+{
+    private string          $loaderIndex;
+    private LoaderInterface $loaderClass;
```
i would prefer to call this $loader. $loaderClass suggests its the class name of the loader, but it is the actual instance.

then again, looking at the rest of the code: we do not use the class anywhere afterwards. how about doing the reflection thing in the constructor and have this field be `private string $loaderClassName`?

dbu commented on src/Exception/Binary/Loader/ChainAttemptNotLoadableException.php:
```diff
@@ -0,0 +1,53 @@
+        return $this->loaderIndex;
+    }
+
+    public function getLoaderClass(): LoaderInterface
```
this is not called from outside the class, i think we can drop the method

dbu commented on src/Exception/Binary/Loader/ChainAttemptNotLoadableException.php:
```diff
@@ -0,0 +1,53 @@
+        return $this->loaderClass;
+    }
+
+    public function getLoaderClassName(): string
```
but this one should stay as is, it gives you the classname of the loader, so i like the name.

if we move reflection to the constructor, this will become a simple getter for `$this->loaderClassName`

dbu reviewed (commented):
sorry to continue to argue about the naming :see_no_evil:

i think we can further simplify the exception class and remove some methods we don't need.
